agent/eventhandler/task_handler.go

```diff
@@ -15,63 +15,96 @@ package eventhandler
+type TaskHandler struct {
+	// submitSemaphore for the number of tasks that may be handled at once
+	submitSemaphore utils.Semaphore
+	// taskToEvents is arn:*eventList map so events may be serialized per task
+	tasksToEvents map[string]*eventList
+	// tasksToEventsLock for locking the map
+	tasksToEventsLock sync.RWMutex
+}
+
+// NewTaskHandler returns a pointer to TaskHandler
+func NewTaskHandler() *TaskHandler {
+	return &TaskHandler{
+		tasksToEvents:   make(map[string]*eventList),
+		submitSemaphore: utils.NewSemaphore(concurrentEventCalls),
+	}
 }
 
 // AddTaskEvent queues up a state change for sending using the given client.
-func AddTaskEvent(change api.TaskStateChange, client api.ECSClient) {
-	addEvent(newSendableTaskEvent(change), client)
+func (handler *TaskHandler) AddTaskEvent(change api.TaskStateChange, client api.ECSClient) {
+	handler.addEvent(newSendableTaskEvent(change), client)
 }
 
 // AddContainerEvent queues up a state change for sending using the given client.
-func AddContainerEvent(change api.ContainerStateChange, client api.ECSClient) {
-	addEvent(newSendableContainerEvent(change), client)
+func (handler *TaskHandler) AddContainerEvent(change api.ContainerStateChange, client api.ECSClient) {
+	handler.addEvent(newSendableContainerEvent(change), client)
 }
 
 // Prepares a given event to be sent by adding it to the handler's appropriate
 // eventList
-func addEvent(change *sendableEvent, client api.ECSClient) {
-	var taskList *eventList
-	var preexisting bool
-	log.Info("Adding event", "change", change)
-
-	// TaskEvents lock scope
-	func() {
-		handler.Lock()
-		defer handler.Unlock()
-
-		taskList, preexisting = handler.taskMap[change.taskArn()]
-		if !preexisting {
-			log.Debug("New event", "change", change)
-			taskList = &eventList{List: list.New(), sending: false}
-			handler.taskMap[change.taskArn()] = taskList
-		}
-	}()
-
-	taskList.Lock()
-	defer taskList.Unlock()
+func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
+
+	seelog.Info("TaskHandler, Adding event: ", change)
+
+	eventsToSubmit := handler.getTaskEventList(change)
+
+	eventsToSubmit.eventListLock.Lock()
+	defer eventsToSubmit.eventListLock.Unlock()
 
 	// Update taskEvent
-	taskList.PushBack(change)
+	eventsToSubmit.events.PushBack(change)
 
-	if !taskList.sending {
-		taskList.sending = true
-		go SubmitTaskEvents(taskList, client)
+	if !eventsToSubmit.sending {
+		eventsToSubmit.sending = true
+		go handler.SubmitTaskEvents(eventsToSubmit, client)
 	}
 }
 
+// getTaskEventList gets the eventList from taskToEvent map, and reduces the
+// scope of the taskToEventsLock to just this function
+func (handler *TaskHandler) getTaskEventList(change *sendableEvent) (eventsToSubmit *eventList) {
+
+	handler.tasksToEventsLock.Lock()
+	defer handler.tasksToEventsLock.Unlock()
+
+	eventsToSubmit, ok := handler.tasksToEvents[change.taskArn()]
+
+	if !ok {
+		seelog.Debug("TaskHandler, New event: ", change)
```

Comment:
I know you copied the line from the old code, but "new event" really isn't super clear.  This is an event arriving for a task that does not currently have any events queued; all the events are "new".  Maybe "collecting events for new task"?